### ISSUE TITLE
Twin Macro (Emotion) Not Working

### ISSUE THREAD

**Issue by lavish-kumar** [saadeghi/daisyui#82]
Recently i tried to use Daisy UI with Emotion and Twin Macro. 
While Bundling CSS in JS with Emotion certain Tailwind variables are not working

Dependency Versioning:
"daisyui": "^1.3.5",

Using an example component (CARD) at the given URL : https://daisyui.com/components/card

Expected Output :
![image](https://user-images.githubusercontent.com/26251693/121878375-43d02600-cd29-11eb-8979-96a69478d6fc.png)

Actual Output
![image](https://user-images.githubusercontent.com/26251693/121878153-01a6e480-cd29-11eb-98b5-41295c36a291.png)

Did some digging and found out that that Background Color and Text Color in the compiled CSS-in-JS has few variables missing
![image](https://user-images.githubusercontent.com/26251693/121878635-94e01a00-cd29-11eb-8a50-3befcf700133.png)

The `--tw-bg-opacity` and `--tw-text-opacity` are missing in the CSS, which intern renders a basic white color to the component as the fallback

Temp Fix : ADD  `bg-opacity-100 text-opacity-100` to Twin Macro

Please Help , This issue is occuring for a number of components

**Comment by saadeghi** [comment]
@lavish-kumar I think it might be a bug from Twin and I'm afraid I can't do anything on my end.  
Just like [Tailwind's original colors](https://tailwindcss.com/docs/background-color), Tailwind plugin colors are expected to render with the opacity variable like this:
```css
.bg-black{
  --tw-bg-opacity: 1;
  background-color: rgba(0, 0, 0, var(--tw-bg-opacity));
}
```
I'm not sure why Twin is not adding `--tw-bg-opacity: 1;` to color utilities because without it, any Tailwind plugin that adds new color (and supports color opacity) will break.

I will keep this issue open until there's a solution.
Can you please open an issue on Twin (and mention this issue) ?

**Comment by lavish-kumar** [comment]
@saadeghi  Thanks for the prompt reply.
_I would report this bug at Twin's Repository_

But Meanwhile I would like to bring some light to the following to observations
**I User the following code to test out a few things :**

- I have tried all the `bg-<color>` utility classes in twin and they seem to produce no valid output 

![image](https://user-images.githubusercontent.com/26251693/121899643-3d01dd00-cd42-11eb-9173-d464c71e0773.png)

- But when i introduce another utility class  for example `btn` from DaisyUi  Here is the output , The colors magically appear, `--tw-opacity` also magically changes to defined = 1.

![image](https://user-images.githubusercontent.com/26251693/121899399-04620380-cd42-11eb-928a-b0373785c616.png)

**Comment by lavish-kumar** [comment]
> @saadeghi Thanks for the prompt reply.
> _I would report this bug at Twin's Repository_
> 
> But Meanwhile I would like to bring some light to the following to observations
> **I User the following code to test out a few things :**
> 
> * I have tried all the `bg-<color>` utility classes in twin and they seem to produce no valid output
> 
> ![image](https://user-images.githubusercontent.com/26251693/121899643-3d01dd00-cd42-11eb-9173-d464c71e0773.png)
> 
> * But when i introduce another utility class  for example `btn` from DaisyUi  Here is the output , The colors magically appear, `--tw-opacity` also magically changes to defined = 1.
> 
> ![image](https://user-images.githubusercontent.com/26251693/121899399-04620380-cd42-11eb-928a-b0373785c616.png)

If it would have been an issue with Twin Macro the other classes like `btn` should also not have defined color properties

**Comment by saadeghi** [comment]
@lavish-kumar Yeah, the style of components are [already built](https://unpkg.com/daisyui@1.3.5/dist/styled.js) using a Tailwind/PostCSS setup, Twin won't need to generate any color utility class for the component classes because it's just a ready-to-use CSS-in-JS object.  
So the colors that were used inside components (for example the background color of `btn-primary`) will include opacity variable indeed.
But daisyUI also [adds](https://github.com/saadeghi/daisyui/blob/master/src/colors/index.js) an [object of color names](https://github.com/saadeghi/daisyui/blob/master/src/colors/colorNames.js) to Tailwind config so we can use these color names on all utility classes (like `bg-primary`, `text-primary`, `border-primary` etc).

**Comment by lavish-kumar** [comment]
@saadeghi  Looks Like the logic for enabling opacity in the following  file [COLORS/index](https://github.com/saadeghi/daisyui/blob/master/src/colors/index.js) is failing.

the 2 `IF` blocks are behaving weirdly. `!== undefined` is not working as expected. 

I Stringfied both the destructured variables and they are as follows:
- `opavityValue` = `"var(--tw-bg-opacity)"`
- `opacityVariable` = `"--tw-bg-opacity"`

as the destructured `opacityValue` is a string it causes a return from the first block as `hsla("<color> / var(--tw-bg-opacity)")`

> **I Implemented the following fix as a patch in my node modules and now it works fine. because the 2nd IF block return `hsla("<color> / var(--tw-bg-opacity,1)")`**



![image](https://user-images.githubusercontent.com/26251693/121949563-38efb280-cd76-11eb-9463-69361324c76d.png)

![image](https://user-images.githubusercontent.com/26251693/121950573-6f79fd00-cd77-11eb-93a2-e70ef42551a1.png)

**Comment by saadeghi** [comment]
@lavish-kumar thanks a lot! 
That function is actually based on [this example](https://github.com/adamwathan/tailwind-css-variable-text-opacity-demo) from Adam Wathan and according to the description:  
> The `opacityValue` argument should be respected first if it exists (the gradient plugin uses this for Safari compatibility when fading to transparent), and the `opacityVariable` argument should be respected next if `opacityValue` is undefined.  

Are you sure removing the first `if` actually fixes this issue? I need to test more to see if this doesn't break anything on different environments and difference utility classes.

Alternatively I can add another file for Twin (like [the one for WindiCSS](https://github.com/saadeghi/daisyui/blob/master/src/colors/windi.js) 🙈) for this slightly different function.

**Comment by lavish-kumar** [comment]
As Adam said thats a general format not the exact format. The documentation on this topic is so scares that a definitive answer is unlikely. 

Also i had another approach mind. In the 1st If Block  as `opacityValue` is a string `hsla(<color> / var(--tw-opacity)`we can manuplate the string directly and append the string with a fallback value in the var so the variable would look like `hsla(<color> / var(--tw-opacity, 1)`.

This is again a hackish way but i believe the Undefined check should be working this is the most optimal way. I think the tailwind update to V2.1 might also be the culprit where the value of these variables could have changed.


You can also refference This Video by [Tailwind Labs Official Channel](https://youtu.be/MAtaT8BZEAo)

From Minute 17 mark.   They also depend on the same logic checking `opacityValue` to be  `undefined` .  But it works for them.



**I am also going to check this out of diffrent environment as i am very keen to get this resolved. I personally love Daisy Ui and am planning to Introduce a extension to this library In JSX format so that this library can reach its heights of reusability**

**Comment by lavish-kumar** [comment]
> @lavish-kumar thanks a lot! 
> That function is actually based on [this example](https://github.com/adamwathan/tailwind-css-variable-text-opacity-demo) from Adam Wathan and according to the description:  
> > The `opacityValue` argument should be respected first if it exists (the gradient plugin uses this for Safari compatibility when fading to transparent), and the `opacityVariable` argument should be respected next if `opacityValue` is undefined.  
> 
> Are you sure removing the first `if` actually fixes this issue? I need to test more to see if this doesn't break anything on different environments and difference utility classes.
> 
> Alternatively I can add another file for Twin (like [the one for WindiCSS](https://github.com/saadeghi/daisyui/blob/master/src/colors/windi.js) 🙈) for this slightly different function.



After a brief code review I think removing 1st block is not safe because if and when opacity is actually specifically defined in style this might fail.
